core/src/main/java/org/apache/calcite/rel/rules/CommonRelSubExprRegisterRule.java

```diff
@@ -0,0 +1,103 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.CommonRelExpressionRegistry;
+import org.apache.calcite.plan.CommonRelSubExprRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Aggregate;
+import org.apache.calcite.rel.core.Filter;
+import org.apache.calcite.rel.core.Join;
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.calcite.rel.core.Project;
+import org.apache.calcite.rel.metadata.RelMetadataQuery;
+
+import com.google.common.collect.Multimap;
+
+import org.immutables.value.Value;
+
+import java.util.function.Predicate;
+
+/**
+ * Rule for saving relational expressions that appear more than once in a query tree to the planner
+ * context.
+ */
+@Value.Enclosing
+public final class CommonRelSubExprRegisterRule extends CommonRelSubExprRule {
+
+  private CommonRelSubExprRegisterRule(Config config) {
+    super(config);
+  }
+
+  @Override public void onMatch(final RelOptRuleCall call) {
+    CommonRelExpressionRegistry r =
+        call.getPlanner().getContext().unwrap(CommonRelExpressionRegistry.class);
+    if (r != null) {
+      r.add(call.rel(0));
+    }
+  }
+
+  /**
+   * A predicate determining if a relational expression is interesting.
+   *
+   * <p>The notion of interesting is loosely defined on purpose since it may change as the API
+   * evolves. At the moment an expression is considered interesting if it contains at least one of
+   * the following RelNode types:
+   * <ul>
+   *   <li>{@link Join}</li>
+   *   <li>{@link Aggregate}</li>
+   *   <li>{@link Filter}</li>
+   * </ul>
+   */
+  private static final class InterestingRelNodePredicate implements Predicate<RelNode> {
+    @Override public boolean test(final RelNode rel) {
+      RelMetadataQuery mq = rel.getCluster().getMetadataQuery();
+      Multimap<Class<? extends RelNode>, RelNode> types = mq.getNodeTypes(rel);
+      if (types == null) {
+        return false;
+      }
+      return types.keySet().stream().anyMatch(
+          t -> Join.class.isAssignableFrom(t) || Aggregate.class.isAssignableFrom(t)
+              || Filter.class.isAssignableFrom(t));
+    }
+  }
+
+  /** Rule configuration. */
+  @Value.Immutable
+  public interface Config extends CommonRelSubExprRule.Config {
+    Config JOIN = ImmutableCommonRelSubExprRegisterRule.Config.builder()
+        .withOperandSupplier(o -> o.operand(Join.class)
+            .predicate(j -> JoinRelType.INNER == j.getJoinType()).anyInputs())
```

Comment:
Does it only support INNER JOIN? Does CALC or SETOP need support?

Reply:
These rules are used by the BasicSuggester which is only meant to be a simple implementation that puts in use the new APIs. If there are reasons to make the implementation more complex we can discuss about it but probably does need to happen at this stage unless you have some concrete use-case in mind.

Note that these rules here match on a specific node (e.g., Join) but there are no restrictions about its descendants. Technically, you can have anything (Union, Minus, Calc, etc.) as long as it is below the join.

Reply:
I wonder if it would be possible to provide users with flexible configurations, such as if I only want to find subtrees rooted in UNION or JOIN. And then give a default configuration, such as searching for AGGREGATE, FILTER, JOIN (inner), and FILTER by default?